### ISSUE TITLE
Stable Gradient for BinaryFocalCrossEntropy

### ISSUE THREAD

**Issue by CV-42** [tensorflow/tensorflow#70326]
### Issue type

Feature Request

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.15.0 and 2.18.0-dev20240624

### Custom code

Yes

### OS platform and distribution

Google Colab

### Python version

3.10.12

### Current behavior?

Hi,

I'm new to writing issues... But here is what I found:

The Gradient-Tape of BinaryFocalCrossEntropy returns nan-values for large inputs. Meanwhile the Gradient-Tape of the normal BinaryCrossEntropy (without "Focal") yields nice values. It would be awesome to have stable gradient calculations for both.

### Standalone code to reproduce the issue

```shell
# Sample data:
y_true = tf.Variable([0, 1, 1])
logits = tf.Variable([-2, .3, 19]) # The "19" is big enough to yield the behavior below.

with tf.GradientTape(persistent=True) as tape:
    # Should be two equivalent definitions due to gamma=0:
    bfce = tf.keras.losses.binary_focal_crossentropy(y_true, logits, from_logits=True, gamma=0)
    bce  = tf.keras.losses.binary_crossentropy(y_true, logits, from_logits=True)

# Values are indeed similar:
print(bfce)
print(bce)
# tf.Tensor(0.22709441, shape=(), dtype=float32)
# tf.Tensor(0.22709441, shape=(), dtype=float32)

# But gradients are not (nan-value):
print(tape.gradient(bce, logits))
print(tape.gradient(bfce, logits))
# tf.Tensor([ 3.9734304e-02 -1.4185250e-01  -1.8675990e-09], shape=(3,), dtype=float32)
# tf.Tensor([ 0.0397343     -0.1418525      nan           ], shape=(3,), dtype=float32)
```

**Comment by CV-42** [comment]
Thanks for the reply!

I had a complicated pipeline where the training produced nan-weights after a few epochs. I could try to find a simplified pipeline producing nans, if you really like. But I believe, that the problem only lies in the way the gradient is calculated. And I think that the problem can be solved without using float64. Here is the explanation:

In the cross-entropy and its gradient, a term of the form
          e^x / (1+e^x) 
 may appear.
 There are different ways of calculating this term:

```
#using float32:
def f32(x):
    return np.exp(np.float32(x)) / (np.float32(1) + np.exp(np.float32(x)))

#using float64:
def f64(x):
    return np.exp(np.float64(x)) / (np.float64(1) + np.exp(np.float64(x)))

# or using float32, but rewriting the term as e^x / (1+e^x) = 1 / (e^(-x)+1):
def f32_stable(x):
    return np.float32(1) / (np.exp(np.float32(-x)) + np.float32(1))

## x=90 is too big for a calculation with float32:
x = 90
print(f32(x))
# output: nan

# This can be solved by using float64:
print(f64(x))
# output: 1.0

# But it can also be solved with rewriting e^x / (1+e^x) = 1 / (e^(-x)+1)
print(f32_stable(x))
# output: 1.0
```

I heard that one can manually provide gradients for a function. And I guess that the normal BinaryCrossEntropy does this.

**Comment by tilakrayal** [comment]
@CV-42,
Apologies for the delay. As the tf.keras.losses.BinaryFocalCrossentropy is related to keras, could you please try to raise the issue in the keras-team/keras repo for the quick resolution. Thank you!
